include/dali/core/endian_util.h

```diff
@@ -0,0 +1,214 @@
+};
+
+#ifdef __linux__
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+/// @brief The host machine is little-endian
+static constexpr bool is_little_endian = true;
+/// @brief The host machine is big endian
+static constexpr bool is_big_endian = false;
+#elif __BYTE_ORDER == __BIG_ENDIAN
+/// @brief The host machine is little-endian
+static constexpr bool is_little_endian = false;
+/// @brief The host machine is big endian
+static constexpr bool is_big_endian = true;
+#else
+#error "Cannot establish endianness of the target system."
+#endif
+#else
+#ifdef __GNUC__
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmultichar"
+#endif
+#if ('ABCD' == 0x41424344UL)
+/// @brief The host machine is little-endian
+static constexpr bool is_little_endian = true;
+/// @brief The host machine is big endian
+static constexpr bool is_big_endian = false;
+#elif ('ABCD' == 0x44434241UL)
+/// @brief The host machine is little-endian
+static constexpr bool is_little_endian = false;
+/// @brief The host machine is big endian
+static constexpr bool is_big_endian = true;
+#else
+#error "Cannot establish endianness of the target system."
+#endif
+#ifdef __GNUC__
+#pragma GCC diagnostic pop
+#endif
+#endif
+
+#define DALI_SWAP_FIELD_ENDIANNESS(unused, unused2, field) swap_endian(s.field);
+
+/**
+ * @brief Used for generating field-wise endianness swapping function for a structure _Struct
+ *
+ * Usage:
+ * ```
+ * struct Size {
+ *   int16_t width, heigth;
+ * };
+ * SWAP_ENDIAN_FIELDS(Size, width, heigth);
+ *
+ * Size s = { 640, 480 };
+ * to_big_endian(s);                // convert to big-endian
+ * fwrite(&s, sizeof(s), 1, file);  // store in a big-endian file
+ * ```
+ *
+ * NOTE: It does not support inheritance - the fields from the base class have to
+ *       be enumerated as well
+ */
+#define SWAP_ENDIAN_FIELDS(_Struct, ...)\
+template <>\
+void swap_endian_impl<_Struct>::do_swap(_Struct &s) {\
+  BOOST_PP_SEQ_FOR_EACH(DALI_SWAP_FIELD_ENDIANNESS, unused, \
+    BOOST_PP_VARIADIC_TO_SEQ(__VA_ARGS__))\
+}\
+
+/**
+ * @brief Converts (in place) the object to little endian
+ *
+ * If the host system is little endian, this function is a no-op.
+ */
+template <typename T>
+void to_little_endian(T &t) {
+  if (is_big_endian)
+    swap_endian(t);
+}
+
+/**
+ * @brief Converts (in place) the object from little endian
+ *
+ * If the host system is little endian, this function is a no-op.
+ */
+template <typename T>
+void from_little_endian(T &t) {
+  if (is_big_endian)
+    swap_endian(t);
+}
+
+/**
+ * @brief Converts (in place) the object to big endian
+ *
+ * If the host system is big endian, this function is a no-op.
+ */
+template <typename T>
+void to_big_endian(T &t) {
+  if (is_little_endian)
+    swap_endian(t);
+}
+
+/**
+ * @brief Converts (in place) the object from big endian
+ *
+ * If the host system is big endian, this function is a no-op.
+ */
+template <typename T>
+void from_big_endian(T &t) {
+  if (is_little_endian)
+    swap_endian(t);
+}
```

Comment:
Can you explain a bit a rationale behind designing this API like this? 

As far as I understand, `from_little_endian` and `to_big_endian` are doing the same thing, but I need to know the device "endianness" to know, which of these I need to choose to work. And if I choose incorrectly, there will be a silent no-op, regardless of the "endianness" of the argument. Is this understanding correct?
It looks like it assumes, that the argument has the "endianness" of the system, but it is not clear to me why. Are there some intangibles I am missing? After all, this functionality enables you to change precisely that.

Reply:
The rationale is as follows:
If you obtain a raw structure in a particular (known and fixed) byte order, you call `from_that_order`.
Conversely, if you have a structure in host order and want to send/write/... it in a fixed byte order, you call `to_that_order`.
Internally, `from_` and `to_` do the same thing - just like POSIX `htonl` and `ntohl`. This wouldn't hold for "middle endian", but I don't think we plan to target [PDP-11](https://en.wikipedia.org/wiki/PDP-11) ;)

Reply:
Ok. I suggest adding this info to the documentation for these functions. Right now, they say "convert in place" but it looks like intention was different.